### PR TITLE
Add query execution plan extensibility APIs

### PR DESCRIPTION
**Pull request by kburtram**
Here are the new extensibility APIs defined in this PR.  I'll add additional APIs to control query options in another commit.

```TypeScript
// existing DMP interfaces
// Notes: currently there are execution options to turn on estimated and actual query plans.
// This is fine for immediate requirements, but we'll probably want to make this more generic
// to support additional options. 
export interface ExecutionPlanOptions {
	displayEstimatedQueryPlan?: boolean;
	displayActualQueryPlan?: boolean;
}

export interface QueryProvider {
	runQuery(ownerUri: string, selection: ISelectionData, runOptions?: ExecutionPlanOptions): Thenable<void>;	
}

/**
	* Namespace for interacting with query editor
*/
export namespace queryeditor {
	export type QueryEvent =
		| 'queryStart'
		| 'queryStop'
		| 'executionPlan';

	export interface QueryEventListener {
		onQueryEvent(type: QueryEvent, document: queryeditor.QueryDocument, args: any);
	}

	// new extensibility interfaces
	export interface QueryDocument {
		providerId: string;

		uri: string;

		// get the document's execution options
		getOptions(): Map<string, string>;

		// set the document's execution options
		setOptions(options: Map<string, string>): void;

		// tab content is build using the modelview UI builder APIs
		// probably should rename DialogTab class since it is useful outside dialogs
		createQueryTab(tab: window.DialogTab): void;
	}

	/**
		* Run query if it is a query editor and it is already opened.
		* @param {string} fileUri file URI for the query editor
		*/
	export function runQuery(fileUri: string, options?: Map<string, string>): void;

	/**
		* Register a query event listener
		*/
	export function registerQueryEventListener(listener: queryeditor.QueryEventListener): void;

	export function getQueryDocument(fileUri: string): queryeditor.QueryDocument
}
``` 

Additionally, we can consider implementing a declarative tab interface similar to what we use in dashboard and some other places.  This is a common pattern that is used through out VS Code.

Declarative extensibility has some advantages, such as it also the host to be more restrictive on the type of extensibility supported and therefore the host can provide more guarantees regarding perf and reliability, etc.  However, it does give up flexibility to the extension.  I think declarative APIs in the ADS use-case should be consider in additional to programmatic APIs, but not in place of the APIs.

Here's an example of what this configuration may look like.
```JavaScript
"query.tabs": [{
	"id": "query-plan-viewer",
	"description": "Query Plan Viewer tab",
	"provider": "MSSQL",
	"title": "Query Plan Viewer",
	"when": "connectionProvider == 'MSSQL' && !mssql:isQueryPlanAvailable",
	"container": {
		"modelview-container": {
			"query-plan-viewer-container": {}
		}
	}
}]
```

Additionally, we want to allow extensions to add toolbar buttons to query editors.  Buttons and menu items are relatively limited in functionality and work well with a declarative extensibility mechanism.  Here's an example of how a toolbar button might be added to a query editor.

```JavaScript
"toolbar": {
	"document/query": [{
		"command": "toggleProcessPlan",
		"when": "connectionProvider == MSSQL",
		"disableWhen": "mssql:iscloud",
		"isToggleButton": true,
		"group": "queryPlan"
	}]
}
```

Here's an example custom query tab from the sample in the PR.

screen shot 2019-02-16 at 4 08 55 pm

### CODE REVIEW

**Comment by kburtram** [comment]
I'm going to use this PR for the query execution plan extensibility work.  This PR isn't ready for merge, but I wanted to publish to get feedback on the general event listener approach.

File: extensions/mssql/src/main.ts [hunk 1]
```diff
@@ -107,6 +107,13 @@ export async function activate(context: vscode.ExtensionContext): Promise<MssqlE
 		credentialsStore.start();
 		resourceProvider.start();
 
+		sqlops.queryeditor.registerQueryInfoListener({
```
Comment (kburtram) [inline extensions/mssql/src/main.ts]: this is an example of the API usage, mssql extension changes won't be merged.

File: extensions/mssql/src/main.ts [hunk 2]
```diff
@@ -190,8 +197,8 @@ async function handleNewNotebookTask(oeContext?: sqlops.ObjectExplorerContext, p
 		// Get the file path after '/HDFS'
 		let hdfsPath: string = oeContext.nodeInfo.nodePath.substring(oeContext.nodeInfo.nodePath.indexOf('/HDFS') + '/HDFS'.length);
 		if (hdfsPath.length > 0) {
-			let analyzeCommand = "#" + msgSampleCodeDataFrame + os.EOL + "df = (spark.read.option(\"inferSchema\", \"true\")"
-				+ os.EOL + ".option(\"header\", \"true\")" + os.EOL + ".csv('{0}'))" + os.EOL + "df.show(10)";
+			let analyzeCommand = '#' + msgSampleCodeDataFrame + os.EOL + 'df = (spark.read.option("inferSchema", "true")'
```
Comment (kburtram) [inline extensions/mssql/src/main.ts]: ...unrelated lint cleanup

File: src/sql/platform/query/common/queryRunner.ts [hunk 3]
```diff
@@ -356,7 +359,15 @@ export default class QueryRunner extends Disposable {
 				// check if this result has show plan, this needs work, it won't work for any other provider
 				let hasShowPlan = !!result.resultSetSummary.columnInfo.find(e => e.columnName === 'Microsoft SQL Server 2005 XML Showplan');
 				if (hasShowPlan) {
-					this.getQueryRows(0, 1, result.resultSetSummary.batchId, result.resultSetSummary.id).then(e => this._planXml.resolve(e.resultSubset.rows[0][0].displayValue));
+					this.getQueryRows(0, 1, result.resultSetSummary.batchId, result.resultSetSummary.id).then(e => {
+						let planXmlString = e.resultSubset.rows[0][0].displayValue;
+						this._eventEmitter.emit(EventType.QUERY_PLAN_AVAILABLE, {
+							providerId: 'MSSQL',
```
Comment (kburtram) [inline src/sql/platform/query/common/queryRunner.ts]: need to find the best way to lookup the active provider ID from this context

**Comment by kburtram** [comment]
Here is another possible API for this functionality.

```TypeScript
// existing DMP interfaces
// Notes: currently there are execution options to turn on estimated and actual query plans.
// This is fine for immediate requirements, but we'll probably want to make this more generic
// to support additional options. 
export interface ExecutionPlanOptions {
	displayEstimatedQueryPlan?: boolean;
	displayActualQueryPlan?: boolean;
}

export interface QueryProvider {
	runQuery(ownerUri: string, selection: ISelectionData, runOptions?: ExecutionPlanOptions): Thenable<void>;	
}

// new extensibility interfaces
export interface QueryDocument {
	providerId: string;

	fileUri: string;

	// tab content is build using the modelview UI builder APIs
	// probably should rename DialogTab class since it is useful outside dialogs
	createQueryTab(tab: window.modelviewdialog.DialogTab): void;
}

export interface QueryInfoListener {
	// instead of having start, stop, planavailable, batchavailable, etc. methods
	// have a single method with a type parameter.
	// another alternative is to have seperate methods for common actions: start, stop, etc. 
	// and then have generic method for misc. events
	onQueryEvent(type: string, editor: QueryDocument);
}

// query editor extensibility namespace
export namespace queryeditor {

	// existing method
	export function runQuery(fileUri: string): void;

	// new methods
	export function runQuery(fileUri: string, options: Map<string, string>): void;

	export function registerQueryInfoListener(listener: QueryInfoListener): void;
}
```

**Comment by anthonydresser** [comment]
Is there a reason we couldn't use vscode menu contribution points for the toolbar?

**Comment by kburtram** [comment]
@anthonydresser Kevin mentioned that too during our review meeting, and I still need to follow-up.  I think the basic reason would be along the lines of a toolbar not being a menu, so it may not make sense to overload the one to be treated as the other (for example, menus can have submenus, which doesn't make sense for toolbars, and perhaps other differences).

Do you know if VS Code is using menu contributions to represent buttons anywhere else?

**Comment by anthonydresser** [comment]
The editor buttons are via menu contributions 
image

I think its `editor/title`

**Comment by kburtram** [comment]
@anthonydresser @kevcunnane if this is what VS Code is doing elsewhere then it probably makes sense to stay consistent.  I'll take a look and update the notes above.  The toolbar support will come in through a separate PR, so there will be another chance to review those details.

**Comment by kevcunnane** [comment]
So - I took a look at your PR and actually think long term toolbar might make sense. Primary reason is menus only work for buttons. We need to extend long term to support drop downs and other non-button features (even sticky / toggle buttons). That might be better done via a new contribution point. Apologies for flip flopping on this one!

File: extensions/queryplan/src/main.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+'use strict';
+
+import * as azdata from 'azdata';
+import * as vscode from 'vscode';
+
+let toggleOn: boolean = false;
+let statusView = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right);
+statusView.text = 'Query Plan watcher on';
```
Comment (alanrenmsft) [inline extensions/queryplan/src/main.ts]: localize?

File: src/sql/azdata.proposed.d.ts [hunk 5]
```diff
@@ -3918,6 +3918,31 @@ declare module 'azdata' {
...
+
+		export interface QueryEventListener {
+			onQueryEvent(type: QueryEvent, document: queryeditor.QueryDocument, args: any);
+		}
+
+		// new extensibility interfaces
+		export interface QueryDocument {
+			providerId: string;
+
+			uri: string;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts]: fileUri?

File: src/sql/azdata.proposed.d.ts [hunk 6]
```diff
@@ -3918,6 +3918,31 @@ declare module 'azdata' {
...
+
+			// get the document's execution options
+			getOptions(): Map<string, string>;
+
+			// set the document's execution options
+			setOptions(options: Map<string, string>): void;
+
+			// tab content is build using the modelview UI builder APIs
+			// probably should rename DialogTab class since it is useful outside dialogs
+			createQueryTab(tab: window.DialogTab): void;
```
Comment (alanrenmsft) [inline src/sql/azdata.proposed.d.ts]: It seems to me AddQueryTab makes more sense here.

File: src/sql/parts/query/editor/queryResultsView.ts [hunk 7]
```diff
@@ -223,6 +226,23 @@ export class QueryResultsView extends Disposable {
 				this._panelView.pushTab(this.topOperationsTab);
 			}
 		}
+
+		// restore query model view tabs
+		this.input.state.visibleTabs.forEach(tabId => {
+			if (tabId.startsWith('querymodelview;')) {
+				// tab id format is 'tab type;title;model view id'
+				let parts = tabId.split(';');
+				if (parts.length === 3) {
```
Comment (alanrenmsft) [inline src/sql/parts/query/editor/queryResultsView.ts]: log an error if it is not 3, otherwise it will be hard to figure out the reason.

File: src/sql/parts/query/editor/queryResultsView.ts [hunk 8]
```diff
@@ -223,6 +226,23 @@ export class QueryResultsView extends Disposable {
...
+					let tab = this._register(new QueryModelViewTab(parts[1], this.instantiationService));
```
Comment (anthonydresser) [inline src/sql/parts/query/editor/queryResultsView.ts]: Are tabs tied to inputs? If so they should be registered in the inputs so they are disposed with the inputs not the results view

File: src/sql/azdata.proposed.d.ts (hunk 5 above)
Comment (kburtram) [reply src/sql/azdata.proposed.d.ts]: it looks like VS Code is using just `uri` in their `TextDocument` so I think that makes more sense for us as well in this case.

File: src/sql/azdata.proposed.d.ts (hunk 6 above)
Comment (kburtram) [reply src/sql/azdata.proposed.d.ts]: `create` seems to be the verb we're using for this type of action generally.  We can discuss whether to change globally in a design review going from proposed to stable, but for now I want to keep consistent with what already in the API.

File: src/sql/parts/query/editor/queryResultsView.ts (hunk 7 above)
Comment (kburtram) [reply src/sql/parts/query/editor/queryResultsView.ts]: we build this string.  it is not user provided.  I don't see a great way to surface the error here and since this isn't user provided, it doesn't seem necessary.

File: src/sql/parts/query/editor/queryResultsView.ts (hunk 8 above)
Comment (kburtram) [reply src/sql/parts/query/editor/queryResultsView.ts]: Tabs are tied to a given resultset, the input's lifetime is longer than tabs.

**Comment by kburtram** [comment]
Let's take this with the current scope of changes and then do the query options and toolbars separately as discussed above.  If I missed any comments please leave them on the next PR.
